### ISSUE TITLE
Extend redirect to include ip not only port

### ISSUE THREAD

**Issue by ChillerDragon** [ddnet/ddnet#7480]
Currently the server can tell a client to reconnect to a different port. I would like to also allow redirection to another ip.

heinrich was concerned about security. I would like to reopen this discussion since moving tees from one host to another is a nice feature for admins and automated orchestration.

**Comment by heinrich5991** [comment]
Perhaps we could try to make sure the target server is informed about this redirect. This way, not a lot of resources are going to be spent on the redirection.

You're still pinging the target server with UDP packets. Perhaps we could require that it's also registered?

**Comment by ChillerDragon** [comment]
> You're still pinging the target server with UDP packets. Perhaps we could require that it's also registered?

Takes away the feature of hiding servers and then redirecting players to it that solved a puzzle. Or hiding servers from ddos requiring the ddoser to join the server before he can attack the target.

**Comment by heinrich5991** [comment]
Do you have a suggestion of how to prevent to maliciously redirect clients?

**Comment by ChillerDragon** [comment]
> Perhaps we could try to make sure the target server is informed about this redirect. This way, not a lot of resources are going to be spent on the redirection.


I am okay with this. But not sure if I understand it or it's use case. Are you thinking informing the server that it might receive redirects? Or asking for permission and only doing it if granted?

A bad actor server can always ignore that protocol? Or are you saying the client should ask the server it is being redirected to?

**Comment by ChillerDragon** [comment]
> Do you have a suggestion of how to prevent to maliciously redirect clients?

No I do not. But I also do not understand what that means. Could you name a specific scenario?

**Comment by ChillerDragon** [comment]
I assume it would be bad to get redirected to some underground government ips that then put you on some sort of blacklist or send the killer squad after you. I assume similar could be achieved already by pinging those ips with spoofed ips.

One possible way to combat that is to include a blacklist on the client side. I think I saw lists already of ips that one should not contact.

**Comment by Learath2** [comment]
I think this can be allowed. 301 redirects on the web are the model I have in mind for this and those are fine. Just need to be careful with sending secrets (f.e. password) to whatever server we get redirected to.

For informing purposes, maybe we can fit a flag somewhere in the connect packet that this is a redirect, so that the receiving servers can outright decline redirects if they don't want them.

**Comment by ChillerDragon** [comment]
> For informing purposes, maybe we can fit a flag somewhere in the connect packet that this is a redirect, so that the receiving servers can outright decline redirects if they don't want them.

I still do not think it is needed but I don't have any problem with doing that. We could use [NET_CTRLMSG_CONNECT](https://chillerdragon.github.io/teeworlds-protocol/06/ctrl_messages.html#NET_CTRLMSG_CONNECT) it has 501 null bytes against reflection attacks I do not think that the server checks if the bytes are zero just the size of the packet. We could fit all kinds of meta data in there including that it is a redirect.

> Just need to be careful with sending secrets (f.e. password) to whatever server we get redirected to.

I would actually propose to allow sending secrets. https://github.com/ddnet/ddnet/issues/8260
If a server redirects a client to another passworded server it would be nice if the client could also receive a password from the source server to login to the target server.


So I propose the following arguments for the net message:

**PORT
[IP]
[PASSWORD]**

**Comment by heinrich5991** [comment]
> For informing purposes, maybe we can fit a flag somewhere in the connect packet that this is a redirect, so that the receiving servers can outright decline redirects if they don't want them.

I think the flag should be opt-in, not opt-out.

**Comment by Learath2** [comment]
> > For informing purposes, maybe we can fit a flag somewhere in the connect packet that this is a redirect, so that the receiving servers can outright decline redirects if they don't want them.
> 
> I think the flag should be opt-in, not opt-out.

Are we talking about the same flag? I think the `IsRedirect` flag we put into the connect packet should be mandatory. `AcceptRedirects` flag on the server idk, I was thinking opt-out but maybe opt-in is fine too if you don't feel comfortable with it. I don't really see much issue to opt-out though

**Comment by heinrich5991** [comment]
Ah, I meant the server side (but did not say so).

**Comment by ChillerDragon** [comment]
What is the point of opt in @heinrich5991 ? How should the information be obtained that the server allows it? By whom? By the connecting client or the redirecting server?

If it requires a network packet to the server to ask if redirect is allowed whats the point. If the server wants no traffic it also wants no traffic asking if traffic is okay. If the server wants no tees then whats the point of the server? In such cases the server admin should just configure a server password or player slots limit.

I still don't understand what dangers you see. What could be the possible harm of unexpected/unwanted redirect? It just makes things more complicated and more annoying to debug for people who want to use the feature.

**Comment by heinrich5991** [comment]
> What is the point of opt in @heinrich5991 ?

The point is that you won't get your server "raided" by someone with a more popular server.

> How should the information be obtained that the server allows it? By whom? By the connecting client or the redirecting server?

By the connecting client.

> I still don't understand what dangers you see. What could be the possible harm of unexpected/unwanted redirect?

See above.

**Comment by ChillerDragon** [comment]
> The point is that you won't get your server "raided" by someone with a more popular server.

In which scenario is that a problem for anyone?

**Comment by heinrich5991** [comment]
> > The point is that you won't get your server "raided" by someone with a more popular server.
> 
> In which scenario is that a problem for anyone?

Someone is being harassed for being a blob. They're hosting a server. A group of blob-haters gather on some other server. The admin of that server finds it funny to redirect everyone to the server of the person who is a blob.

**Comment by ChillerDragon** [comment]
> Someone is being harassed for being a blob. They're hosting a server. A group of blob-haters gather on some other server. The admin of that server finds it funny to redirect everyone to the server of the person who is a blob.

That is not realistic in my head.

- Is blob harassing actually a thing? I might be blind but I have not seen it in tw.
- The admin would empty his own server. That is usually not desirable for an admin. I don't think "funny" is worth it.
- The blob haters could also directly join the blob server.
- I assume most blob haters would prefer to keep playing on the server they intentionally joined so they probably leave the blob server quickly.

Do you know of any harassment case that happend in the last decade where this would have been applicable? Imo the problem here is the blob harassment and not the redirect. I don't see how the redirect feature is making it worse. Or am I missing a crucial point here? How is that scenario a problem that gets worse with redirects?

How did you think of that example? Is that a problem elsewhere? In twitch streaming maybe?

**Comment by ChillerDragon** [comment]
I am still waiting for a response here. My opinion stands I think it is a good idea and I am waiting for more details on why it is not a good idea. I am ready to implement the code if the proposal is accepted. Sorry for bumping but this was stale since may and I am actively waiting for this proposal to get accepted.

Also I wanted to mention that the popular game minecraft just seemed have added [transfer packets](https://wiki.vg/Protocol#Transfer_.28configuration.29). If it makes anyone feel better we can wait a bit and see what went wrong for them. Their transfers are opt in by default.

```
accepts-transfers 	boolean 	default: false 	Allows servers to accept incoming transfers via a transfer packet.
							    false - Incoming transfers are rejected, and the player is disconnected.
							    true - Incoming transfers are allowed, and the server must approve it.
```
[source](https://minecraft.wiki/w/Server.properties)

**Comment by Sedonya** [comment]
Maybe only allow IP redirection to servers that have a community?

**Comment by heinrich5991** [comment]
> opt in by default

That's called opt-out.

**Comment by jxsl13** [comment]
@heinrich5991 I highly agree on that one

**Comment by ChillerDragon** [comment]
Another feature might be hub servers that then redirect players to a server hosted in their ping region. I often see new players playing on ddnet servers with 200 ping which gives them a bad first impression of how laggy the game is.

Also @heinrich5991 I am still waiting for a response from you to my [questions](https://github.com/ddnet/ddnet/issues/7480#issuecomment-2094070094).

**Comment by def-** [comment]
This probably doesn't really belong here, but I've also noticed new players on wrong server, here are my thoughts:

It would already help to show accurate ping for players, currently we guess it based on geolocation, which is inherently inexact. We could indeed do this:

1. DDNet master server runs a traceroute to the client IP, picks the IP 1-2 stops before the client to prevent leaking the actual player IP
2. Each IP of game servers runs a ping server that the DDNet master server contacts with the new IP to ping
3. The ping server returns the pings to the DDNet master server which returns it to the client (probably quite delayed, so show the out-of-date/geolocation-based data in the meantime)

This should give us pretty good ping times (the route could be different than from client to server, and we are missing 1-2 hops, so not perfect) without revealing actual player IPs to any server hosting, keeping things fair for every game server hoster (instead of special-casing DDNet/community servers).

Easier solution would be to just have a toggle with which servers to ping directly: DDNet, community, all, and give players a choice in the beginning, defaulted to DDNet only so they at least get good Tutorial+Novice experience and because they have to trust DDNet anyway to run the client.

**Comment by ChillerDragon** [comment]
@def- you should probably open an new issue for that. Otherwise it will get lost I assume. But yes as someone who is hosting servers without a community I would very much appreciate getting ranked more precisely by ping.

I was more talking about cases likes these.

![image](https://github.com/user-attachments/assets/a7fc95de-e071-46f1-8691-c194740971e1)

This is not a ddos. I tried multiple times telling the players that they are probably on the wrong server. I don't think any of them got my messages. These users probably would not look at the ping in the server browser.

Tbh I don't really see official ddnet servers using the redirect feature to solve this. But I would totally manually move users on my servers when I am unable to communicate in the chat.

**Comment by ChillerDragon** [comment]
I do not fully understand the problems, so I am having a hard time coming up with solutions for it. But if this feature is blocked by hosters receiving unwanted new players from redirects we could introduce some request/accept protocol. I don’t like the complexity of it but it still allows me to use the feature for everything I wanted to use it for. A redirect could look like this:

- server A asks client foo to connect to server B
- client foo drops from server A. Then it sends a message to server B saying it was redirected from server A and waits for a response
- server B responds with redirect accept/deny based on the sv_allow_redirects config (which by default is 0)
- Only if the client foo receives a accept from B it will proceed to connect. Otherwise it drops from server B and returns to the server browser.

This would solve unexpected raids (still no idea how this is even a problem). And because the redirect request contains the source ip it allows to later add black and whitelists for servers you accept redirects from.

To avoid reflection attacks the join requests needs to be big enough. Which would increase traffic bandwidth compared to a naive redirect. It’s not nice but should be fine.

I really think this would be a cool feature. It also allows rebooting/upgrading a VPS without users having to stop their sessions. The players and their game state could be sent to another server before reboot.

**Comment by ChillerDragon** [comment]
@LW-Davide made the valid point that redirects can be abused to farm ips. Malicious actors can fake popular servers and register master server entries that redirect to the real servers. This way it might happen that players never notice what is happening and the servers do not get reported. Then the malicious actor can get access to the ip addresses of the players which is bad. The same thing can already be done without redirects but then it is more obvious to the players that something is wrong.

This attack surface can be reduced if all servers drop redirects from external ips by default. Then only the servers where the admins intentionally opt in to use the feature can be affected. Which should be a way lower player count which might not contain the attackers desired ips. Another option would be the annoying popup for the user to click "OK I WANT TO BE REDIRECTED TO ANOTHER IP" but I do not think that is good. It can scare users if the feature is used in a legit way. And it will basically ban all afk players. Also it tends to get annoying and users might still click on it without questioning it which is totally fair because it might also be a legitimate redirect. Depending on the connection speed users should see two connection screens with different ips either way.

**Comment by sjrc6** [comment]
I support adding this feature, I don't think the stated security risks outweigh the benefits and seem poorly justified. 

But if we would rather be careful I think the simplest solution is the client should include the server address that it's coming from in the connection request to the new server in some way (and letting it know it was redirected, inherently).

This alone would let servers implement their own whitelist or blacklist filtering for redirects however they want, since they can just drop the client if they do not want to accept redirects from that address. DDNet could add an official implementation of this this but that is optional and can be implemented later if and of the stated concerns actually happen. 

I don't believe that a complex protocol where the client or server asks permission to redirect is needed before attempting it, it should be the responsibility of the server doing the redirect to know the policy of the receiving server, if anything giving advanced permission before dropping the client only makes abuse easier.

**Comment by ChillerDragon** [comment]
> the client should include the server address that it's coming from in the connection request to the new server in some way (and letting it know it was redirected, inherently).

I had the same thought and like it too. It can be spoofed by the client but the redirecting server can not spoof it for the client so it could be of some value.

> IMO giving advanced permission before dropping the client only makes abuse easier since you can try to "raid" many servers without dropping any clients.

I was thinking rejected clients end up in the main menu not in the original server.

**Comment by heinrich5991** [comment]
> I don't believe that a complex protocol where the client or server asks permission to redirect is needed, it should be the responsibility of the server doing the redirect to know the policy of the receiving server. IMO giving advanced permission before dropping the client only makes abuse easier since you can try to "raid" many servers without dropping any clients.

I don't think the protocol has to be complex. Let me suggest a simple one:

The server, before sending the initial map change message, sends a list of allowed IP addresses that the client may be redirected from. The client compares that to the IP address it was redirected from. If the list contains the IP address, it continues to connect. If it doesn't, it disconnects and shows a message to the user. After the user confirms the message, they're back to the serverbrowser.

**Comment by heinrich5991** [comment]
Maybe it'd be even easier if this "allowed IP" list was in the serverbrowser list. This way, we don't even need to connect to the target server to figure out whether they want our traffic or not. This wouldn't allow to redirect to servers on a different IP address that aren't in the masterserver list though.

**Comment by ChillerDragon** [comment]
> This wouldn't allow to redirect to servers on a different IP address that aren't in the masterserver list though.

That is not good. I want to hide servers from the master server and only redirect clients I verified there.

**Comment by heinrich5991** [comment]
> That is not good. I want to hide servers from the master server and only redirect clients I verified there.

Can you tell me more about your use case? Is it about hiding the IP address of the server you're redirecting to?

(If it's that you probably need to be careful: we leak the IP addresses of the server that is connected-to via the steam integration.)

**Comment by ChillerDragon** [comment]
The use case is not making the server a target for ddosers. Anything not listed in the master is way less attractive for griefers.

**Comment by heinrich5991** [comment]
I see, so some security through obscurity. I guess we can try to support that.

The first solution I proposed works for that.

**Comment by ChillerDragon** [comment]
If the list of ips gets too big it can be used for reflection attacks potentially.

**Comment by ChillerDragon** [comment]
> The server, before sending the initial map change message, sends a list of allowed IP addresses that the client may be redirected from. The client compares that to the IP address it was redirected from. If the list contains the IP address, it continues to connect. If it doesn't, it disconnects and shows a message to the user. After the user confirms the message, they're back to the serverbrowser.

I do not like that the server has to expose its list and that the client can decide. Also as I server I want to know the source server even if I allow redirects from everywhere. Imo the client should send the ip and the server can then decide to approve it or not.

**Comment by heinrich5991** [comment]
> I do not like that the server has to expose its list and that the client can decide.

This is also the way [CORS headers](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing) work for HTTPS, it seems to work well for them.

> Also as server I want to know the source server even if I allow redirects from everywhere.

That's similar to an [HTTP `Referer`](https://en.wikipedia.org/wiki/HTTP_referer) header and I don't think the new server owner should get more data than they technically need.

**Comment by ChillerDragon** [comment]
> That's similar to an [HTTP Referer](https://en.wikipedia.org/wiki/HTTP_referer) header and I don't think the new server owner should get more data than they technically need.

If I allow all how do I know who to start blocking once bad traffic comes in (still don't know what bad traffic is tbh but just for the sake of argument).
I want to know who to thank. I would be super curious what happend when a wall of players connected. If it looks the same as a regular connection on the network level and the source can not be traced then it will be harder to differentiate between redirects and flood attacks.

**Comment by sjrc6** [comment]
> I had the same thought and like it too. It can be spoofed by the client but the redirecting server can not spoof it for the client so it could be of some value.

I don't see how the client spoofing this is ever a threat because a malicious client can just do a regular connection request to any server and disobey the redirect protocol entirely since they control the whole client. So we are only talking about things possible on the official client. 

> Can you tell me more about your use case? Is it about hiding the IP address of the server you're redirecting to?

It could be potentially very useful for ddos protection in a scenario where you let clients connect to a public master-registered server and then add them to a firewall whitelist then redirect them to a private server, the public server forwards the player data to the master server, so you do not ever need to expose the traffic of the real server to authorized requests.

**Comment by sjrc6** [comment]
> I was thinking rejected clients end up in the main menu not in the original server.

I don't think I implied the rejected clients should end up on the original server, they are at the mercy of the redirect server after the server sends them there. So if they're dropped they end up in the server browser or wherever menu they go. 

> I don't think the protocol has to be complex. Let me suggest a simple one:
> 
> The server, before sending the initial map change message, sends a list of allowed IP addresses that the client may be redirected from. The client compares that to the IP address it was redirected from. If the list contains the IP address, it continues to connect. If it doesn't, it disconnects and shows a message to the user. After the user confirms the message, they're back to the serverbrowser.

Is there a issue with the protocol I proposed? I think it's the most elegant from both the client and server perspective. If you disallow ip redirect by default then I don't see anything that can go wrong.

**Comment by heinrich5991** [comment]
> Is there any issue with the protocol I proposed? I think it's the most elegant from both the client and server perspective. If you disallow ip redirect by default then I don't see anything that can go wrong. If ddnet doesn't want this feature they can just leave it off on their servers.

The issue is that it becomes the easiest way to just accept redirects from anywhere. By building the redirect whitelist into the protocol, we avoid that people choose "from anywhere" by default.

**Comment by sjrc6** [comment]
> The issue is that it becomes the easiest way to just accept redirects from anywhere. By building the redirect whitelist into the protocol, we avoid that people choose "from anywhere" by default.

You can build whatever whitelist system you want as part of the server, I really can't see a scenario where abuse becomes an issue. 

DDNet, KoG and any server network who leaves the settings at default will not accept redirects from "anywhere" so that automatically removes 90+% of abuse targets. If a smaller network wants to use redirects they can just add their few IPs to the provided server whitelist implementation.

Having the ability to accept directs from anywhere seems fine, the client reporting the IP allows the server to rate limit redirects from a single source, or just rate limit them in general if they are so afraid of "raiding" 

Reporting the IP also makes it easy for servers to blacklist redirect sources they don't like which further limits abuse.

I think the chance that anyone pursues abusing the system in the first place is low and it's so trivial to react to abuse if it does happen that I can't understand what your issue with this. 

Everything that is possible with this system is already possible if you just tell the player to copy paste the "connect 0.0.0.0" message that you send in in chat into their console.

**Comment by heinrich5991** [comment]
Let me ask the opposite question: What kind of thing do you want to build that would not work under my proposed protocol?

I gave some reasons for why the referer should maybe not be shown and why it's cool to have the whitelist built into the protocol so that the easiest way to implement the protocol isn't some sort of "allow all" policy.

**Comment by sjrc6** [comment]
> Let me ask the opposite question: What kind of thing do you want to build that would not work under my proposed protocol? 

It seems more complex than necessary and will therefore not be implemented by anyone in the near future.

I also see no reason to disallow redirects from any server. As someone who does not believe in the likelihood that this will be abused allowing all directs to my server would be free player traffic.

**Comment by heinrich5991** [comment]
> It seems more complex than necessary and will therefore not be implemented by anyone in the near future.

I don't think the level of complexity is high. See https://github.com/ddnet/ddnet/issues/7480#issuecomment-2653094588. One message that gets added. If someone has a use case for the redirects, they can implement it.

If nobody has a use case for the redirects, then it's better if they're not implemented because it's code that we'll have to support going forward.

**Comment by ChillerDragon** [comment]
> If nobody has a use case for the redirects, then it's better if they're not implemented because it's code that we'll have to support going forward.

I have 3 servers with 3 different ips and I want to be able to move my users between them. I need this very much this is why opened this issue.

@heinrich5991 so https://github.com/ddnet/ddnet/issues/7480#issuecomment-2653094588 is an accepted implementation I can send a pr for? Then we can remove the needs-discussion label.

**Comment by ChillerDragon** [comment]
I think the origin check should happen client and server side so it needs malicious clients and server to bypass it but whatever we can also only do it client sided.

**Comment by heinrich5991** [comment]
I think you can send a PR for https://github.com/ddnet/ddnet/issues/7480#issuecomment-2653094588.

**Comment by ChillerDragon** [comment]
I AM GOING TO SEND PR THIS WINTER 🌨️ 🚀
